sdcm/gemini_thread.py

```diff
@@ -0,0 +1,187 @@
+                        event.add_info_and_publish(node=self.node, line=d['message'], line_number=line_number)
+                if self.stopped():
+                    break
+            if self.stopped():
+                break
+
+
+class GeminiStressThread(object):
+
+    def __init__(self, test_cluster, oracle_cluster, loaders, gemini_cmd, timeout=None, outputdir=None):
+        self.loaders = loaders
+        self.gemini_cmd = gemini_cmd
+        self.test_cluster = test_cluster
+        self.oracle_cluster = oracle_cluster
+        self.timeout = timeout
+        self.result_futures = []
+        self.gemini_log = None
+        self.outputdir = outputdir
+
+    def _generate_gemini_command(self, loader_idx):
+        seed = random.randint(1, 100)
+        test_node = random.choice(self.test_cluster.nodes)
+        oracle_node = random.choice(self.oracle_cluster.nodes)
+        self.gemini_log = '/tmp/gemini-l{}-{}.log'.format(loader_idx, uuid.uuid4())
+        cmd = "/$HOME/{} --test-cluster={} --oracle-cluster={} --outfile {} --seed {}".format(self.gemini_cmd.strip(),
+                                                                                              test_node.ip_address,
+                                                                                              oracle_node.ip_address,
+                                                                                              self.gemini_log,
+                                                                                              seed)
+        return cmd
+
+    def run(self):
+
+        executor = concurrent.futures.ThreadPoolExecutor(max_workers=len(self.loaders.nodes))
+        for loader_idx, loader in enumerate(self.loaders.nodes):
+            self.result_futures.append(executor.submit(self._run_gemini, loader, loader_idx))
+
+        return self
+
+    def _run_gemini(self, node, loader_idx):
+        logdir = os.path.join(self.outputdir, node.name)
+        try:
+            os.makedirs(logdir)
+        except OSError:
+            pass
+        log_file_name = os.path.join(logdir,
+                                     'gemini-l%s-%s.log' %
+                                     (loader_idx, uuid.uuid4()))
+        gemini_cmd = self._generate_gemini_command(loader_idx)
+
+        GeminiToolEvent(type='start', cmd=gemini_cmd)
+        try:
+            with GeminiToolEventsPublisher(node=node, gemini_log_filename=log_file_name):
+
+                result = node.remoter.run(cmd=gemini_cmd,
+                                          timeout=self.timeout,
+                                          ignore_status=False,
+                                          log_file=log_file_name)
+                # sleep to gather all latest log messages
+                time.sleep(5)
+        except Exception as details:
+            LOGGER.error(details)
+            result = details.result
```

Comment:
can be more Exception here would rather pick the result with `result = getattr(details, 'resault', 'n/a')` or something like that

Reply:
fixed